### PR TITLE
Add support for other PushID overloads

### PR DESCRIPTION
**Pull request by michaelfairley**
Or, a different approach: https://github.com/michaelfairley/imgui-rs/tree/push_id_overloads2

### CODE REVIEW

**Comment by Gekkio** [comment]
Thanks!

I think I prefer the trait version in theory, but there's some problems with it:

* The only way to implement it correctly is to use unsafe code -> pretty much makes the trait pointless from an implementation point of view
* It violates all lifetime and initialization guarantees of `imgui-rs`. I can call `"lol".push()` at any time with very bad consequences

What if we made an enum instead and supported converting to it? Something like

```
pub enum ImId<'a> {
  Int(i32), Str(&'a str), ImStr(&'a ImStr), Ptr(*const c_void)
}
```

Then you could:

* implement `Into<ImId<'static>> for i32` and others
* accept `<I: Into<ImId<'a>>` in `push_id` and `with_id`
* match based on the enum and call the right push function in push_id -> dangerous pushing is limited to one place where we already do it and have certain guarantees

So, in other words, right now the polymorphism in the API is all about pushing (= a side-effect), but it could be better if it was just about knowing which of the supported id types is being used.

Also, there are some conflicts that need to be resolved.

**Comment by michaelfairley** [comment]
Great points. This resulted in a complete rewrite, so I've opened #133 to have cleaner history.
